src/Titanium.Web.Proxy/Certificates/CertificateManager.cs

```diff
@@ -481,29 +485,48 @@ private X509Certificate2 makeCertificate(string certificateName, bool isRootCert
                 return cached.Certificate;
             }
 
-            // handle burst requests with same certificate name
-            // by checking for existing task for same certificate name
-            if (pendingCertificateCreationTasks.TryGetValue(certificateName, out var task))
-            {
-                return await task;
-            }
-
-            // run certificate creation task & add it to pending tasks
-            task = Task.Run(() =>
-            {
-                var result = CreateCertificate(certificateName, false);
-                if (result != null)
+            Task<X509Certificate2?> createCertificateTask;
+            await pendingCertificateCreationTaskLock.WaitAsync();
+            try {
+                // handle burst requests with same certificate name
+                // by checking for existing task for same certificate name
+                if (pendingCertificateCreationTasks.TryGetValue(certificateName, out createCertificateTask))
                 {
-                    cachedCertificates.TryAdd(certificateName, new CachedCertificate(result));
+                    return await createCertificateTask;
```

Comment:
If we have 25 requests for 5 different hosts, this would mean all remaining 4 hosts will have to wait here for other host certificate generation. I will take a look at the issue later today. But this seem to be not a good solution.

Reply:
the code block inside the WaitAsync/Release should be fast-ish (creates a task) and is only hit prior to the certificate creation.